### DISCUSSION TITLE
本地包的相关路径应该具有更高的优先级

### DISCUSSION THREAD

**Discussion by q962** [xmake-io/xmake#5432] [Q&A]
我安装了本地包，结果还是链接了系统的文件

**Comment by q962** [comment]
我想使用  xmake f --linkdirs 先跳过这个错误。

结果没有改变，对比前后，路径的顺序没有改变

**Comment by star-hengxing** [comment]
https://xmake.io/#/package/remote_package?id=disable-system-library

**Reply by q962** [reply]
即使我已经安装了本地包？
`xmake q -y --extra={system=false}`

**Reply by star-hengxing** [reply]
xmake 包默认先找系统库，如果没有才下载包构建。这个选项主要用于禁用 windows(vcpkg)/linux(apt/pacman)/mac(brew) 的 fetch

**Reply by q962** [reply]
```
finding harfbuzz from xmake ..
checking for xmake::harfbuzz ... no
finding harfbuzz from vcpkg ..
finding harfbuzz from conan ..
finding harfbuzz from pkgconfig ..
checking for harfbuzz ... pkgconfig::harfbuzz 9.0.0
finding libpng from xmake ..
checking for xmake::libpng ... no
finding libpng from pacman ..
checking for pacman::libpng ... libpng 1.6.43
```

但是输出结果第一个是 xmake，这个结果和你的结论冲突

**Reply by q962** [reply]
我的标题就证明了结论。
因为本地库的lib路径已经出现在编译命令里。
这说明已经优先使用了本地库。

**Reply by star-hengxing** [reply]
你确定在 add requires 都设置 system = false 了么

**Reply by q962** [reply]
我只为我需要的库设置了  system = false。
是的，我尝试。
和我执行` xmake q -y --extra={system=false}` 命令后的编译命令是一致的

**Reply by q962** [reply]
```
...objs
-m64  \
-LD:/msys64/mingw64/lib  \
-LD:/msys64/mingw64/lib/../lib  \
-LC:/Users/user/AppData/Roaming/.xmake/packages/l/lua/v5.4.7/1faf1df8f0dc4f9aaa2007626bb208eb/lib  \
-LD:/msys64/mingw64/bin/../lib  \
...
```

**Reply by star-hengxing** [reply]
已经用系统包管理安装了 libpng，无法复现。

```lua
add_requires("libpng", {system = false})
```

```bash
$ mingw
$ xmake f -c -v -p mingw
checking for architecture ... x86_64
checking for mingw directory ... C:/Users/star/scoop/apps/msys2/2024-07-27/mingw64
checking for unzip ... no
checking for 7z ... ok
checking for git ... ok
checking for gzip ... ok
checking for tar ... ok
git rev-parse HEAD
checking for xmake::zlib ... no
checking for pacman ... ok
checking for pkg-config ... no
checking for pkgconf ... no
checking for cygpath ... ok
checking for pacman::zlib ... zlib 1.3.1
checking for xmake::libpng ... no
note: install or modify (m) these packages (pass -y to skip confirm)?
in xmake-repo:
  -> libpng v1.6.43
please input: y (y/n/m)
```

system = true

```sh
$ xmake f -c -v -p mingw
checking for architecture ... x86_64
checking for mingw directory ... C:/Users/star/scoop/apps/msys2/2024-07-27/mingw64
checking for unzip ... no
checking for 7z ... ok
checking for git ... ok
checking for gzip ... ok
checking for tar ... ok
git rev-parse HEAD
checking for xmake::zlib ... no
checking for pacman ... ok
checking for pkg-config ... no
checking for pkgconf ... no
checking for cygpath ... ok
checking for pacman::zlib ... zlib 1.3.1
checking for xmake::libpng ... no
checking for pacman::libpng ... libpng 1.6.43
```

**Reply by q962** [reply]
你注意看我的上面发的编译命令。那是 xmake 的编译命令
很明显的，在本地包的lib路径之前有系统包路径。

**Comment by star-hengxing** [comment]
才想起以前提过类似的问题 https://github.com/xmake-io/xmake/issues/5066
需要改进 xmake 对系统包的查找才行，或者不混用系统包和 xmake 包

**Reply by star-hengxing** [reply]
可以追踪 https://github.com/xmake-io/xmake/issues/5433 ，我稍后看看怎么实现

**Reply by q962** [reply]
代码太庞大了。看的我头晕😵
哪里存在介绍项目组成的介绍吗？

我认为不是查找问题，目前的查找顺序是对的，没有出现错误。
现在的错误是，`-L` 和 `-I` 不跟随包的查找顺序。或者说本地具有更高的优先级

我跟踪了一下
```
-- in core/project/target.lua
target:orderpkgs()
    return   packages.push(  ...self:get("packages", opt)  )
```
这里是出问题的一个地方。`orderpkgs` 函数或者`self:get` 并没有针对优先级的排序

**Reply by star-hengxing** [reply]
怎么判断哪个是系统包路径呢，难道硬编码么，所以只能考虑传库的绝对路径而不是 linkdris + links 的方式。
目前这个 feature，需要拿到库的绝对路径，也就是 libfiles，不过这个信息 xmake 目前没处理。
你可以看下这里，print 一下 result 的内容，然后看看可以怎么设置 result.libfiles，这样才能用 `add_packages("libpng", {linkpath = true})`。

https://github.com/xmake-io/xmake/blob/cc2a7d08129c785ec0344bf76f504b75510c9413/xmake/modules/package/manager/pacman/find_package.lua#L134-L186

**Reply by star-hengxing** [reply]
> 哪里存在介绍项目组成的介绍吗？

一个早期的文章可以看看
https://tboox.org/cn/2017/09/28/xmake-sourcecode-arch/

**Reply by q962** [reply]
linkpath 的作用是什么？

> 怎么判断哪个是系统包路径呢

应该确认一共有几种包的类型
应该只有两种吧
一种是通过非xmake的包管理能够获取到的
另一种就是 xmake 包管理系统本地安装的

所以需要一个字段用于描述这个是属于本地的。或者用一个字段描述这个包的来源。

**Reply by q962** [reply]
在文件 /modules/lib/detect/find_package.lua 124 行，为缓存结果增加 `result.from = found_manager_name`

**Reply by star-hengxing** [reply]
> linkpath 的作用是什么？

这个 issue 有讲 https://github.com/xmake-io/xmake/issues/5066 。。

> 所以需要一个字段用于描述这个是属于本地的。或者用一个字段描述这个包的来源。

能确保系统包路径传对，链接器就会优先搜索么。而且链接器也不止 ld，msvc linker/lld/mold 也是这样处理么。目前传绝对路径 msvc linker 和 ld 起码是有效的。

> 在文件 /modules/lib/detect/find_package.lua 124 行，为缓存结果增加 result.from = found_manager_name

看这个干啥呢，`xmake f -c` 缓存就没了，查找要看我上面的代码

**Reply by q962** [reply]
> 能确保系统包路径传对，链接器就会优先搜索么。而且链接器也不止 ld，msvc linker/lld/mold 也是这样处理么。目前传绝对路径 msvc linker 和 ld 起码是有效的。

没明白，你的意思是，及时将本地包的路径放在最前面，有的编译系统也不会按照参数的顺序去寻找？

> 看这个干啥呢，xmake f -c 缓存就没了，查找要看我上面的代码

存在这样的一个调用链，最上面的这个他会缓存结果啊。
```
 modules/lib/detect/find_package.lua
   modules/package/manager/find_package.lua
       manager/package/manager/*/find_package.lua
```
https://github.com/xmake-io/xmake/blob/cc2a7d08129c785ec0344bf76f504b75510c9413/xmake/modules/lib/detect/find_package.lua#L103-L124

缓存的作用不就是避免重复任务吗，清楚缓存了。表示会以同样的逻辑生成新的缓存文件，还是能知道这个包来自哪里。

**Reply by q962** [reply]
问题是它这个文件缓存文件是 detelt，而不是 packages，应该将结果写入 packages 。
我还没找到相关的代码

**Reply by star-hengxing** [reply]
> 没明白，你的意思是，及时将本地包的路径放在最前面，有的编译系统也不会按照参数的顺序去寻找？

对啊，怎么确保链接器行为一致呢，有什么文档说明么，还是 StackOverflow 有讲。

> 存在这样的一个调用链，最上面的这个他会缓存结果啊。

现在解决这个问题的方式有两个：
- 你说的提高路径优先级，让链接器用 package 的 links 而不是系统的 links。

> 所以需要一个字段用于描述这个是属于本地的。或者用一个字段描述这个包的来源。

现在没有这个机制，你看缓存能干啥。。

- 使用库的绝对路径，直接传给链接器 c:x/y/z.lib。

xmake 已经实现了，不过系统包管理没去适配而已。所以不如来改善这个

**Reply by q962** [reply]
> 现在没有这个机制，你看缓存能干啥。。

确定这个库是xmake安装的库，用于在代码中判断提升路径优先级

> 对啊，怎么确保链接器行为一致呢，有什么文档说明么，还是 StackOverflow 有讲。

msvn / gcc / clang 都会按顺序搜索指定的目录。
至于其他的编译系统，这个差异应该由 xmake 解决。
无非就是控制生成目标的文件名称，提供一个独立无二的名称。
让编译系统不会找错文件。

使用库的绝对路径，这是一个特性和当前的问题没有关系。
你提供了 lib 文件的绝对路径，是否还需要提供 include 的绝对路径？

**Reply by star-hengxing** [reply]
> 使用库的绝对路径，这是一个特性和当前的问题没有关系。
> 你提供了 lib 文件的绝对路径，是否还需要提供 include 的绝对路径？

用绝对路径直接干掉 linkdirs + links 的形式，也就没有你这个问题，且更容易实现

**Reply by q962** [reply]
参数爆炸，现在的参数长度已经够离谱了。

在Windows上，使用 meson 编译 gtk
meson 会生成 ninja，ninja 直接无法运行，参数爆炸
里面全是绝对路径，就离谱。

使用绝对路径是很容的绝对问题。但如果文件很多呢？参数得有多长？
不同的系统对参数的个数是有限制的吧
[windows 是 8191](https://learn.microsoft.com/zh-cn/troubleshoot/windows-client/shell-experience/command-line-string-limitation)
linux 应该没有这个问题

**Reply by q962** [reply]
我目前的参数已经达到了 5399，差不多 30个文件吧。

当让这个可以进行优化，将多个 obj 进行合并。这个工作由 xmake 做？

**Reply by star-hengxing** [reply]
参数太长，自然会优化成用文件传递。这个机制早就有了。
不过还会有 Windows 长路径限制的问题，平时很少出现，暂时无解、

**Reply by waruqi** [reply]
> 我目前的参数已经达到了 5399，差不多 30个文件吧。
> 
> 当让这个可以进行优化，将多个 obj 进行合并。这个工作由 xmake 做？

目前原本就是自动处理了这个问题，否则你参数超过 1024 基本就已经跪了。

**Reply by waruqi** [reply]
> 不过还会有 Windows 长路径限制的问题，平时很少出现，暂时无解、

走 win 的 xmake 安装包，安装时候，就会默认帮用户修改注册表，开启长路径绕过的限制。当然用户也可以选择不开。

**Comment by waruqi** [comment]
这主要是由于 add_packages 即使按顺序追加包 linkdirs links 。。但是其中某个包如果用了 系统库，通过 pkg-config 或者其他系统库查找方式，获取带入了 `-LD:/msys64/mingw64/lib` 等系统根搜索路径后，穿插在 link flags 中间导致搜索顺序被干扰

即使 add_requires 后面的非系统包，xmake 也确实用了 自己安装的版本，`-Lxx` 到了自己的安装路径，但是被前面某个系统包引入的 `-LD:/msys64/mingw64/lib` 干扰了，链接器就会找错库位置。

这种只能说 xmake 本身的搜索行为并没问题，仅仅是包用了 `-L/-l` 链接库的一个副作用。。除非是所有的包改成 `/xxxx/xxx.lib` 完整库文件链接，才能完全规避这个问题。。

也就是通过上面说的 linkpath 方式，让 add_packages 完全走全路径链接，避免被搜索路径干扰。

**Reply by q962** [reply]
> 即使 add_requires 后面的非系统包，xmake 也确实用了 自己安装的版本，-Lxx 到了自己的安装路径，但是被前面某个系统包引入的 -LD:/msys64/mingw64/lib 干扰了，链接器就会找错库位置。

这就是我目前的问题

> 这种只能说 xmake 本身的搜索行为并没问题，仅仅是包用了 -L/-l 链接库的一个副作用。。除非是所有的包改成 /xxxx/xxx.lib 完整库文件链接，才能完全规避这个问题。。

是的，搜索没有问题。本地包的上一个是系统包，导致 mingw64/lib 提前出现，最后还是使用了系统的lib文件。
我不清楚本地lib放在最前面会导致什么问题。
-L 只是修改lib的搜索路径
 -l 的顺序错乱才会出现符号丢失问题。

**Reply by waruqi** [reply]
> -L 只是修改lib的搜索路径
-l 的顺序错乱才会出现符号丢失问题。

不，主要是 -L 影响，目前 add_packages 走 `-L/-l` 只能做到，`-Lxxx -L/mingw64/lib -Lxxx -lxxx -lxxx -lxxxx` 全部 `-L` 和 全部 `-l` 分离的，所以前面一堆 `-L` 就会干扰后面的 `-l` ，如果你中间穿插一个 `-L/mingw64/lib` 在 xmake 包的前面，就会优先从 mingw64 里面找。。

`-l` 的顺序不会错的，除非 包本身内部 links 顺序没配对，或者 add_packages 顺序不对。

所以根本解法目前支持的就是对冲突的包，启用 linkpath 全路径配置，或者全部禁用系统库，不带入 `/mingw64/lib` 。没其他解法。。你配置下 linkpath 就完事了。

**Reply by q962** [reply]
我将本地包作为 add_packages 的第一项，手动提升优先级。

除了手动修改 xmake 文件，会提供其他方式吗？

举个例子：例如在 ci/cd 中运行 xmake
已知目标系统的包版本配置不符合项目需求
所以，我需要为其专门编写一个 build.sh or xmake.lua 去处理包的参数问题。
有可能为了避免麻烦会采用以下的方案
```lua
add_packages("xxx", {
    linkpath = os.getenv("xxx_linkpath") and true or false},
    system  = os.getenv("xxx_system") and true or false},
)
```
会不会发生这种情况。

**Reply by waruqi** [reply]
add_packages 不支持 system 参数。。那个是设置到 `add_requires` 中的。。

ci/cd 上，你全部强制走远程拉取不就好了么。。system = false 全部设置成 false，有全局 policy 可以切 `package.install_only`，也可以 xmake.lua 里面 `add_requireconfs("**", {system = false})` 全部改成非系统库。

**Reply by q962** [reply]
例子举错了, ci/cd 就是应该全部使用非系统库。

但是这种情况会发生。在 docker/Flatpak/Snap/ 或者在什么包管理系统中。或者别的人在自己的环境里编译
在这些情况下，修改 xmake 文件就不是一个很好的选择。
所以需要提供一个方式来影响 xmake 的逻辑。

我不知道 `xmake q -y --extra={system=false,linkpath=true} name`  这会不会修改 xmake的逻辑。

如果可以那就没问题。

**Reply by waruqi** [reply]
> 修改 xmake 文件就不是一个很好的选择。
所以需要提供一个方式来影响 xmake 的逻辑。

不是上面说了，有全局 policy 可以切 package.install_only 么。。policy 不都可以 `xmake f --policies=` 配置么

**Reply by q962** [reply]
不需要全部的包都走远程啊。
只有一个包有问题为什么要导致全部包重新编译呢

**Reply by waruqi** [reply]
那就单独那个包配置 linkpath 。。或者单独那个包 配置 system = false 。。但是命令行参数，不可能单独对某个包。。只能全局改。。

**Reply by q962** [reply]
所以我才会提出那种写法，手写全部的配置的 getenv，以应对可能的错误。
但是那样太繁琐了，也可以 on_load 里来处理这些环境变量。
所以我的建议是 
改善 `xmake q` 命令，将 extra 的结果保存起来。
限制就是，如果在项目目录执行，并且该项目依赖这个包。

要不就 xmake 处理这些环境变量。例如 XMAKE_REQUIRE_xxx_system=1

**Reply by waruqi** [reply]
不考虑。太繁琐，要么你自定义个 option 去控制
